File: CombinationSum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution 
{
private:
	vector<vector<int> > res;
	void helper(vector<int> c, int target, vector<int> &candidates, int s)
	{
		if (target == 0)
		{
			res.push_back(c);
			return;
		}
		if (target < 0)
			return;
		for (int i = s; i < candidates.size(); i++)
		{
			if (i > 0 && candidates[i] == candidates[i - 1])
				continue;
			c.push_back(candidates[i]);
			helper(c, target - candidates[i], candidates, i);
			c.pop_back();
		}
	}
public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) 
    {
    	vector<int> t;
    	sort(candidates.begin(), candidates.end());
    	helper(t, target, candidates, 0);
    	return res;
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/CombinationSum.java
/**
 * Created by xyunpeng on 2/15/16.
 */
public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        Arrays.sort(candidates);
        helper(candidates, target, new ArrayList<Integer>(), 0, result);
        return result;
    }

    private void helper(int[] candidates, int target, List<Integer> now, int sum, List<List<Integer>> result) {
        if (sum > target) {
            return ;
        } else if (sum == target) {
            result.add(new ArrayList(now));
            return ;
        } else {
            for (int i = 0; i < candidates.length; i++) {
                if (now.size() > 0 && candidates[i] < now.get(now.size() - 1)) continue;
                now.add(candidates[i]);
                sum += candidates[i];
                helper(candidates, target, now, sum, result);
                now.remove(now.size() - 1);
                sum -= candidates[i];
            }
        }
    }
}
